File: matlab/clipping.m
function [ clippedVoice ] = clipping( voice, category )
%CLIPPING Three level clipping for pitch frequency estimation
%   input: voice data
%   input: category - '3level' or 'center'
%          '3level' means three level clipping 三电平削波
%          'center' means center clipping 中心削波
%   output: clipped voiced data

    L=length(voice);
    m = max(voice);
    n = min(voice);

    coeff = 0.68;
    threshold = coeff*m;
    
    a = voice;
    
    if category == '3level'
        for k=1:L ;                          %clipping
            if a(k)>=threshold
                %a(k)=a(k)-th0;
                a(k)=1;
            elseif a(k)<=(-threshold);
                %a(k)=a(k)+th0;
                a(k)=-1;
            else
                a(k)=0;
            end      
        end
    else
        for k=1:L ;                          %clipping
            if a(k)>=threshold
                a(k)=a(k)-threshold;
                %a(k)=1;
            elseif a(k)<=(-threshold);
                a(k)=a(k)+threshold;
                %a(k)=-1;
            else
                a(k)=0;
            end      
        end
    end
        
    clippedVoice = a;
    
end

File: Monitor_c/clipping.c
#include "clipping.h"
#include "max.h"
#include <malloc.h>

int clipping(double data[], int size, int category, double **clippedData)
{
	int max_pos;
	double maximum = max(data,size,1,-1,&max_pos);
	double coeff = 0.68;
	double threshold = coeff*maximum;
	int i,k;
	double *clipped = (double *)malloc(sizeof(double)*size);

	if(category == 1)	// 3 level clipping
		for(i=0; i<size; i++)
			if(data[i]>=threshold)
				clipped[i] = 1;
			else if(data[i]<=0.00-threshold)
				clipped[i] = -1;
			else
				clipped[i] = 0;
	else if(category == 2)	//center clipping
		for(i=0; i<size; i++)
			if(data[i]>=threshold)
				clipped[i]=data[i]-threshold;
			else if(data[i]<=0.00-threshold)
				clipped[i]=data[i]+threshold;
			else
				clipped[i]=0;
	else
		return -1;
	
	*clippedData = clipped;
}
